feat: enhance util classes with advanced banking, magic, and walker f…

Voxsylvae opened the pull request:
# Rs2* Utility Classes Changes Summary

## Overview

Changes made to Rs2* utility classes 
## Modified Files

1. **Rs2Bank.java** - Added bank data caching system and enhanced pathfinding
2. **Rs2BankData.java** - NEW: Bank data storage class  
3. **Rs2ItemModel.java** - Added lazy loading for cached items
4. **Rs2Magic.java** - Added rune calculation methods
5. **Rs2Spellbook.java** - Added spellbook switching functionality
6. **Rs2Random.java** - Added normal distribution methods
7. **Rs2PrayerEnum.java** - Updated to use VarbitID constants
8. **Rs2Walker.java** - Added transport analysis and banking routes

## Changes

### Plugin Scheduler Requirements System

#### PathfinderConfig

- **Enhanced transport validation**: Added support for bank items in transport currency checking
- Transports can now use items from the bank for currency requirements when `useBankItems` is enabled
- **Improved currency validation logic**: Modified to check both inventory and bank for required transport currencies

### Bank Data Caching System

#### Rs2Bank

- Added persistent bank data storage using RuneLite config
- Bank items survive client restarts
- Separate cache for each account profile
- Added methods: `loadInitialBankStateFromConfig()`, `saveBankToConfig()`, `hasCachedBankData()`
- Enhanced `getNearestBank()` with performance improvements and better bank filtering
- Added `getPathToNearestBank()` methods for complete pathfinding to banks
- Improved accessibility checking with performance tracking
- **Latest Updates:**
  - Added performance measurements and logging for bank search operations
  -Not done yet: Optimized expensive `hasRequirements()` filtering operation with performance tracking, have no solution for it 

#### Rs2BankData (New Class)

- Stores bank items as `[id, quantity, slot]` arrays for serialization
- Lazy rebuilding of Rs2ItemModel objects when accessed
- Methods: `getBankItems()`, `size()`, `isEmpty()`, `set()`, `setEmpty()`

#### Rs2ItemModel

- Added `createFromCache(id, quantity, slot)` factory method
- ItemComposition loaded only when properties are accessed

### Magic System

#### Rs2Magic

- Added `getRequiredRunes(Rs2Spells spell, int casts, boolean checkRunePouch)`
- Added `getRequiredRunes(Rs2CombatSpells spell, int casts, boolean checkRunePouch)`
- Considers equipped staves and tomes
- Checks inventory and rune pouch for available runes

#### Rs2Spellbook

- Added spellbook switching locations and quest requirements
- Added `switchTo()` method for automatic spellbook changing
- Added `getCurrentSpellbook()` and unlock checking methods
- Added `getNearestSwitchLocation()` with pathfinding integration

### Walker Transport System

#### Rs2Walker

- Added `getTransportsForDestination()` - analyzes required transport items
- Added `compareRoutes()` - compares direct vs banking routes
- Added `walkWithBankedTransports()` - automatic banking for transport items
- Added `findNearestAccessibleTarget()` - pathfinding to nearest target from list
- Integration with Rs2Magic for teleportation spell rune requirements
- **Latest Updates:**
  - Enhanced `getTransportsForPath()` with improved filtering and requirement setup
  - Added comprehensive route analysis with performance metrics in `compareRoutes()`
  - Implemented intelligent banking workflows with `walkWithBankingState()`
  - Added transport validation for currency-based transports and spell requirements
  - Improved transport filtering with `applyTransportFiltering()` for fairy rings and currency items

### Other Changes

#### Rs2Random

- Added `fancyNormalSample()` for human-like random delays
- Added `truncatedNormalSample()` for bounded normal distribution

#### Rs2PrayerEnum

- Replaced deprecated `Varbits` constants with `VarbitID` constants
- Updated 5 prayer enums: THICK_SKIN, BURST_STRENGTH, CLARITY_THOUGHT, SHARP_EYE, MYSTIC_WILL

## Improvements

- Bank data loads from cache without opening bank
- ItemComposition loaded only when needed
- Plugins can check bank contents without opening bank
- Automatic transport item analysis for destinations
- Automatic spellbook switching for spells
- Banking vs direct route comparison
- Bank data persists across client restarts
- Separate data for different account types

## Breaking Changes

- `storeBankItemsInMemory()` renamed to `updateLocalBank()`
- Direct access to `bankItems` field should use `bankItems()` method

## New Public Methods

### Rs2Bank New Methods

- `loadInitialBankStateFromConfig()`
- `saveBankToConfig()`
- `hasCachedBankData()`
- `emptyBankState()`

### Rs2Magic New Methods

- `getRequiredRunes(Rs2Spells, int, boolean)`
- `getRequiredRunes(Rs2CombatSpells, int, boolean)`

### Rs2Spellbook New Methods

- `getCurrentSpellbook()`
- `switchTo()`
- `isLunarSpellBookUnlocked()`
- `isAncientSpellBookUnlocked()`
- `getNearestSwitchLocation()`

### Rs2Walker New Methods

- `getTransportsForDestination()`
- `compareRoutes()`
- `walkWithBankedTransports()`
- `findNearestAccessibleTarget()`

### Rs2Random New Methods

- `fancyNormalSample(double, double)`
- `fancyNormalSample(int, int)`
- `truncatedNormalSample(double, double, Double, Double)`

Voxsylvae committed:
feat: enhance util classes with advanced banking, magic, and walker functionality

- Rs2Bank: Add caching, add utility for path to nearest bank
- Rs2BankData: Implement bank data management system
- Rs2Magic/Rs2Spellbook: improved spellbook utility class, added functions for getting runes for Rs2Spell, considering equiped staff and tomes
- Rs2Walker: Advanced waking function, with travel analysis, comparing route with banked items, added utility function walkWithBankedTransports
	- added other transport related utility functions, getTransportsForDestination, getTransportsForDestination
	- compareRoutes-> Compares the efficiency of traveling directly to a target versus going via bank first and use banked items
	- fix bug: interaction for shiping boat and pay-fare, dialogue handling
- Rs2Prayer: Improved prayer management
- Rs2ItemModel: Enhanced inventory modeling, need for bank caching
- Rs2Random: Mathematical utility functions

feat: improve bank stander script functionality
- better error handling
- Rs2bank changes integration
feat: enhance shortest path plugin with improved configuration and pathfinding
- Enhanced configuration options for debuging
- add pathfinder configuration consider banked  Currency in the bank when useBankedItems flags is set

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/walker/TransportRouteAnalysis.java:
```diff
@@ -0,0 +1,67 @@
+     * Analyzes whether it's more efficient to travel directly to a destination or
+     * first visit a bank to obtain required transport items.
+     */
+public  class TransportRouteAnalysis {
```
Can we use Lombok here?

Voxsylvae replied:
Done

Voxsylvae committed:
refactor: clean up debug logging and improve bank pathfinding organization

Features:
- Extract bank pathfinding logic into reusable helper method getPathAndBankToNearestBank()
- Improve return type to provide both path and bank location in single call
- Add proper requirements validation for object-based bank selection
- Simplify getNearestBank() and getPathToNearestBank() method implementations

Cleanup:
- Remove debug log statement from Rs2Bank.hasBankItem() method
- Remove excessive logging from Rs2Walker.getTotalTilesFromPath()
- Clean up verbose logging in Rs2Walker performance analysis
- Standardize log output formatting with consistent indentation
- Remove redundant pathfinding code duplication

Performance:
- Reduce method complexity by extracting common pathfinding logic
- Eliminate duplicate bank search operations
- Streamline path calculation workflows

Technical Changes:
- 2 files modified: Rs2Bank.java, Rs2Walker.java
- Net reduction: 62 lines of code
- Maintains backward compatibility for all public methods

Voxsylvae committed:
fixes, cleanup, to walkWithBankingState

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/prayer/Rs2Prayer.java:
```diff
@@ -195,4 +196,28 @@ public static Rs2PrayerEnum getBestMeleePrayer() {
 
         return null;
     }
+    public static boolean rigourUnlocked() {
```
I think we should adjust the naming of these methods to `isRigourUnlocked()`, `isPietyUnlocked()` etc

Voxsylvae replied:
Done

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/inventory/Rs2ItemModel.java:
```diff
@@ -65,6 +59,123 @@ public Rs2ItemModel(Item item, ItemComposition itemComposition, int slot) {
+                    this.isTradeable = itemComposition.isTradeable();
+                }
+                this.inventoryActions = itemComposition.getInventoryActions();
+                addEquipmentActions(itemComposition);
```
addEquipmentActions does require the client thread - can we wrap this within Microbot.getClientThread().runOnClientThreadOptional() ?

Voxsylvae replied:
I don't think it's necessary because we're already getting the itemComposition from the client thread, and within addEquipmentActions we're only manipulating things from the Rs2ItemModel class.
Am I missing something that would require the client thread here? 🙈

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/prayer/Rs2PrayerEnum.java:
```diff
@@ -2,18 +2,19 @@
-    CLARITY_THOUGHT("Clarity of Thought", 35454987, 7, Varbits.PRAYER_CLARITY_OF_THOUGHT, 2),
-    SHARP_EYE("Sharp Eye", 35455003, 8, Varbits.PRAYER_SHARP_EYE, 3),
-    MYSTIC_WILL("Mystic Will", 35455006, 9, Varbits.PRAYER_MYSTIC_WILL, 19),
+    THICK_SKIN("Thick Skin", 35454985, 1, VarbitID.PRAYER_THICKSKIN, 0),
```
I am not sure if it makes sense to only do this for a few of the prayers, should we just migrate all of them into the new gameval.VarbitID class?

Voxsylvae replied:
Done

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/walker/Rs2Walker.java:
```diff
@@ -2284,4 +2505,772 @@ public static boolean isTeleportItem(int itemId) {
+            List<Transport> transports = getTransportsForPath(path, 0, prefTransportType, true);
+            
+            // Log found transports for debugging
+            transports.forEach(t -> log.info("Transport found: " + t));
```
If these are meant to be debug logs - lets use log.debug

Voxsylvae replied:
done

gmason0 reviewed (commented):
Thank you for taking the time to create comments & create a beautiful features. Most of my requests are nit-picks based on what I've found in the review, but otherwise this is great - thanks!

Voxsylvae committed:
feat: enhance TransportRouteAnalysis with comprehensive utility methods and improve code consistency

- Add extensive utility methods to TransportRouteAnalysis for transport path analysis:
  * getTransportsForDirectPath/BankingPath with customizable parameters
  * getMissingTransports methods for different path segments
  * Transport item quantity analysis methods
  * Proper Lombok @Getter annotations and comprehensive JavaDoc

- Improve Rs2Prayer method naming consistency:
  * Add 'is' prefix to boolean methods (rigourUnlocked -> isRigourUnlocked)
  * Fix typo: perservereUnlocked -> isPreserveUnlocked

- Standardize Rs2PrayerEnum to use VarbitID constants instead of Varbits
- Remove unused import from Rs2PrayerEnum
- Reduce log noise: change transport logging from info to debug level

These enhancements prepare for future optimization of walkWithBankedTransportsAndState
by enabling path reuse instead of recalculation, while maintaining current functionality.

gmason0 reviewed (approved):
Thank you Vox for your contributions! You're not only empowering the API but making a difference for everyone that uses these new features <3
